src/matcher/type/date/DateMatcher.ts

```diff
@@ -0,0 +1,41 @@
+import {Matcher} from "../Matcher";
+
+export abstract class DateMatcher extends Matcher {
+    protected readonly date: Date;
+
+    constructor(date: MaybeDate) {
+        super();
+
+        this.date = toDate(date);
+
+        if (!isValidDate(this.date)) {
+            throw new InvalidDateError(date);
+        }
+    }
+}
+
+export class InvalidDateError extends Error {
+    public readonly name = this.constructor.name;
+
+    constructor(date: MaybeDate) {
+        super(`Invalid date: ${date}`);
+
+        Object.setPrototypeOf(this, InvalidDateError.prototype);
+    }
+}
+
+export type MaybeDate = number | string | Date | { seconds: number };
+
+/** @internal */
+export function toDate(maybeDate: MaybeDate): Date {
+    if (typeof (maybeDate as any)?.seconds === "number") {
+        return new Date((maybeDate as { seconds: number }).seconds * 1000);
+    }
+
+    return new Date(maybeDate as number);
```

Comment:
I would prefer not to use any casts, for example like this:

```suggestion
    if (maybeDate instanceof Date) {
      return maybeDate;
    } else if (typeof maybeDate === "object") {
        return new Date(maybeDate.seconds * 1000);
    } else {
      return new Date(maybeDate);
    }
```